### PR TITLE
Analytical Cp Delta: thin-airfoil physics baseline for SRF correction

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current SRF heads predict raw pressure corrections. For the aft-foil specifically, the head must learn both the baseline pressure distribution shape (suction peak, stagnation, recovery) AND the wake-induced perturbation. By subtracting an analytical thin-airfoil pressure baseline (Cp ≈ -2·dY/dX), the SRF head only needs to learn the **delta** — which is smaller in magnitude and concentrated in wake-interaction zones.

This is the DeltaPhi principle (residual learning on physics baseline) applied at the surface level. Critically different from:
- `--residual_prediction` (predicts correction to p=0 freestream — no geometry info)
- Panel Cp (#2169, FAILED) which ran a full panel solver at inference time (slow, noisy, tandem crashes)

The thin-airfoil dY/dX is purely geometric (zero cost), closed-form, and applies per-node.

**Source:** Classical thin-airfoil theory (Lighthill 1951; Abbott & von Doenhoff textbook). DeltaPhi-NO principle for physics-informed residual learning.

**Confidence:** MEDIUM-HIGH. Changes the prediction task (historically the only category that produces wins).

### Step 1: Add config flag

\`\`\`python
analytical_cp_delta: bool = False   # Thin-airfoil Cp baseline subtraction for SRF heads
\`\`\`

### Step 2: Compute thin-airfoil Cp baseline

In \`Transolver.forward()\`, after the main trunk prediction but **before** the SRF heads are applied, compute a per-node thin-airfoil pressure baseline for surface nodes:

\`\`\`python
if cfg.analytical_cp_delta:
    # Use raw (pre-standardized) coordinates to estimate surface tangent dY/dX
    # x_raw is available from the input before standardization
    # For surface nodes: approximate dY/dX from the DSDF gradient direction
    # DSDF1 gradient (features 4:6) gives unit normal for foil-1
    # DSDF2 gradient (features 6:8) gives unit normal for foil-2 (aft-foil)
    # Surface tangent is perpendicular to surface normal
    
    # For foil-1 (fore) surface nodes (boundary_id 5 or 6):
    dsdf1_nx = x_raw[:, :, 4]  # normal x-component
    dsdf1_ny = x_raw[:, :, 5]  # normal y-component
    # Thin-airfoil: Cp ≈ -2 * sin(theta) ≈ -2 * (-nx) for small angles
    cp_baseline_fore = 2.0 * dsdf1_nx  # [B, N], sign: positive for suction side
    
    # For foil-2 (aft) surface nodes (boundary_id 7):
    dsdf2_nx = x_raw[:, :, 6]
    dsdf2_ny = x_raw[:, :, 7]
    cp_baseline_aft = 2.0 * dsdf2_nx
    
    # Clamp to physical range to prevent TE/LE singularity
    cp_baseline_fore = cp_baseline_fore.clamp(-3.0, 1.5).detach()
    cp_baseline_aft = cp_baseline_aft.clamp(-3.0, 1.5).detach()
\`\`\`

### Step 3: Modify SRF head prediction to predict delta

In the surface refinement head application (both standard SRF and aft-foil SRF):

\`\`\`python
# For the standard surface_refine head (fore-foil surface nodes):
if cfg.analytical_cp_delta:
    # SRF head predicts delta_p (correction to analytical baseline)
    # Add the analytical baseline to the pressure channel only
    fore_mask = (boundary_id == 5) | (boundary_id == 6)  # [B, N]
    pred[:, :, 2][fore_mask] = cp_baseline_fore[fore_mask] + srf_correction[:, :, 2][fore_mask]

# For the aft-foil SRF head:
if cfg.analytical_cp_delta and cfg.aft_foil_srf:
    aft_mask = (boundary_id == 7)
    aft_pred[:, :, 2][aft_mask] = cp_baseline_aft[aft_mask] + aft_srf_correction[:, :, 2][aft_mask]
\`\`\`

**IMPORTANT implementation notes:**
1. Use \`.detach()\` on cp_baseline — no gradient through analytical term
2. cp_baseline uses **raw** (pre-standardized) DSDF gradient features. If features are already standardized, compute from raw input
3. Apply ONLY to pressure channel (index 2 of 3-output SRF head: Ux, Uy, p)
4. The \`asinh_pressure\` transform is applied AFTER — Cp baseline in same units as raw pressure prediction
5. Verify DSDF gradient indices (4:6 for foil-1, 6:8 for foil-2) match actual layout in \`tandem_dataset.py\`

### Step 4: Run experiments

**Run 2 seeds:**
\`\`\`bash
cd cfd_tandemfoil && python train.py --agent askeladd --seed 42 \
  --wandb_name "askeladd/analytical-cp-delta-s42" --wandb_group "askeladd/analytical-cp-delta" \
  --analytical_cp_delta \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
\`\`\`

Seed 73: identical but \`--seed 73 --wandb_name "askeladd/analytical-cp-delta-s73"\`.

### Step 5: Diagnostic check first

Run a 2-epoch debug to verify:
1. Cp baseline values look reasonable (histogram: centered ~0, range [-3, 1.5])
2. SRF delta predictions are smaller than without baseline
3. No NaN/Inf
4. torch.compile compatibility

## Baseline

Current best (PR #2184, DCT Freq-Weighted Loss, 2-seed avg):

| Metric | Baseline | Target |
|--------|----------|--------|
| p_in   | 13.205   | < 13.205 |
| p_tan  | **28.502** | **< 28.502** |
| p_oodc | 7.816    | < 7.816 |
| p_re   | 6.453    | < 6.453 |

W&B runs: 6yfv5lio (seed 42, p_tan=28.432), etepxvjc (seed 73, p_tan=28.572).

**Reproduce baseline:**
\`\`\`bash
cd cfd_tandemfoil && python train.py --agent askeladd --wandb_name "askeladd/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
\`\`\`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd: Found implementation issue with DSDF feature interpretation.

The PR instructions state to use `x_raw[:, :, 4]` as the foil-1 surface normal x-component. After inspecting the dataset, I found that x[:, :, 4:12] (the DSDF channels) contain multi-scale **distance values** in range [0, 5.0], NOT unit normal components.

Surface node DSDF stats (from actual data):
- ch0: [0.002, 5.0] mean=3.19
- ch1: [0.011, 5.0] mean=3.37
- ch4: [0.750, 5.0] mean=3.78

These are clamped SDF values, not gradients. Using them as-is gives `cp_baseline ≈ 2 * 3.3 = 6.6` — completely wrong.

**Proceeding with an alternative physics-based baseline**: Using node y-position sign + AoA to approximate upper/lower surface pressure asymmetry:
- Upper surface (y > 0): Cp_base = -2 * AoA (suction)
- Lower surface (y < 0): Cp_base = +2 * AoA (pressure)

This captures the dominant AoA-driven Cp asymmetry from thin-airfoil theory. It's simpler than the DSDF approach but still physics-motivated.

Proceeding with this implementation and running experiments.

**Comment by tcapelle** [comment]
test

**Comment by tcapelle** [comment]
## Experiment Results: Analytical Cp Delta

### Implementation Note
The advisor's PR instructions assumed DSDF features at `x[:,:,4:6]` were surface normal components. Investigation revealed they are **multi-scale clamped SDF distances** (range 0-5), not unit normals. I replaced the approach with a thin-airfoil AoA-based baseline:

```
Cp ≈ -2α × sign(y)
```
- Upper surface (y>0): Cp ≈ -2α (suction)
- Lower surface (y<0): Cp ≈ +2α (pressure)
- `α = AoA0_rad` (fore-foil) or `AoA1_rad` (aft-foil)

### Results (2 seeds)

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed avg | Delta |
|--------|----------|---------|---------|------------|-------|
| p_in   | 13.205   | 13.4    | 14.3    | 13.85      | +4.9% worse |
| p_tan  | **28.502** | **27.8** | **28.8** | **28.30** | **-0.7% better** |
| p_oodc | 7.816    | 11.9    | 10.9    | 11.40      | +45.9% worse |
| p_re   | 6.453    | 8.8     | 8.4     | 8.60       | +33.3% worse |

**W&B runs:** e7lucein (seed 42), vuxgzmui (seed 73)
**W&B group:** askeladd/analytical-cp-delta

### Analysis

**Negative overall.** While p_tan improved slightly (-0.7%), p_oodc and p_re regressed dramatically (+46%, +33%).

**Root cause:** The simplified `Cp ≈ -2α × sign(y)` baseline is too crude:
1. It assumes symmetric airfoils (NACA profiles have camber)
2. It ignores thickness effects and local curvature
3. For OOD conditions (different Re, different AoA ranges), the baseline mismatch grows
4. The SRF head must now learn to correct both the baseline error AND the actual physics delta, making the task harder not easier

**Why p_tan improved slightly:** Tandem transfer involves similar AoA range to training but different foil configurations. The crude baseline still captures the correct sign of the pressure asymmetry.

**Why p_oodc/p_re regressed:** OOD conditions/Re involve flow regimes where Cp ≈ -2α is a poor approximation (e.g., separated flow, high Re effects). SRF capacity wasted correcting baseline error.

### Suggested follow-ups
1. **Close this approach** — the simplified thin-airfoil baseline lacks sufficient accuracy without actual surface normals
2. If surface normals become available in the data pipeline, revisit with proper Cp ≈ 2·nx formulation
3. Consider a **learned per-node baseline** (small MLP predicting baseline from node features) rather than analytical formula

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2212 because ADVISOR: Closing this experiment — clear dead end.

**Results vs current baseline (PR #2207, TE coord frame):**

| Metric | Baseline | PR #2212 avg | Delta |
|--------|----------|-------------|-------|
| p_in   | 12.490   | 13.85       | +10.9% worse |
| p_tan  | 28.521   | 28.30       | -0.8% better |
| p_oodc | 7.618    | 11.40       | +49.7% worse |
| p_re   | 6.411    | 8.60        | +34.2% worse |

While p_tan improved marginally, the massive regressions on p_oodc (+50%) and p_re (+34%) make this a clear negative.

**Key insights to carry forward:**
1. DSDF features at indices 4:12 are multi-scale clamped SDF distances, NOT surface normal components. Good catch.
2. The simplified AoA-based thin-airfoil baseline (Cp ≈ -2α × sign(y)) is too crude — it ignores camber, thickness, curvature, and separation effects.
3. The SRF head is better off learning the full prediction than correcting a noisy baseline.

**Note:** The experiment was also run against the prior baseline (pre-TE coord frame), which further hurts the comparison. The current codebase has `--te_coord_frame` which wasn't included in this run.

Your suggestion about a learned per-node baseline is interesting but would be a fundamentally different experiment (learned vs analytical). I'll assign you a fresh hypothesis shortly. Good work identifying the DSDF feature issue..
